### PR TITLE
child_process helper: fix stderr blocking for discard case. fix #2609

### LINKED ISSUE #2609: Writing to stderr from out_exec_filter plugin, blocks
**Describe the bug**

I'm using `exec_filter` to process some data using Python. My initial command looks like

```
<match logs.**>
    # https://docs.fluentd.org/output/exec_filter
    @type exec_filter

    command /path/to/python/script

    <format>
      @type json
    </format>
    <parse>
      @type json
    </parse>
</match>
```

For logging, my script is using the standard python3 logging library for debugging.
I noticed the script backing up and spent a while trying to debug it, but I now suspect the lower level `child_process_execute` in `fluentd/lib/fluent/plugin_helper/child_process.rb` to be suspect

Updating the above command to `command /path/to/python/script 2>> /var/log/td-agent/script.log` seems to suddenly unblock processing

**Expected behavior**

Within, out_exec_filter, a child process is called via 

```
execute_child_process = ->(index){
    child_process_execute("out_exec_filter_child#{index}".to_sym, @command, on_exit_callback: exit_callback) do |readio, writeio|
        child_process_callback.call(index, readio, writeio)
    end
}
```

which means the parameter defaults to `stderr: :discard` which then arrives at this block

```
if !mode.include?(:stderr) && !mode.include?(:read_with_stderr) && stderr != :discard # connect
    writeio, readio, wait_thread = *Open3.popen2(*spawn_args, spawn_opts)
elsif mode.include?(:read_with_stderr)
    writeio, readio, wait_thread = *Open3.popen2e(*spawn_args, spawn_opts)
else
    writeio, readio, stderrio, wait_thread = *Open3.popen3(*spawn_args, spawn_opts)
end
```

While I have not debugged any further than this, I would expect if fluentd's plugin is not handling stderr, then it would at least default to sending it to /dev/null or some other place. There was 

**Your Environment**

```
fluentd --version 
fluentd 1.7.0
python -m platform
Linux-3.10.0-957.21.3.el7.x86_64-x86_64-with-centos-7.6.1810-Core
```

### PR DESCRIPTION
**Pull request by repeatedly**
Signed-off-by: Masahiro Nakagawa <repeatedly@gmail.com>

**Which issue(s) this PR fixes**: 
Fixes #2609 

**What this PR does / why we need it**: 
reopen with IO::NULL is not correct way to discard message.
It replaces ruby side IO object, not process. Thanks nurse-san!
Need to specify IO::NULL when launch the process.

**Docs Changes**:
No need

**Release Note**: 
Same as title

### CODE REVIEW

File: test/plugin_helper/test_child_process.rb [hunk 1]
```diff
@@ -124,6 +124,39 @@ def configure(conf)
     assert_equal expected, ary
   end
 
+  test 'can execute external command at just once, which can handle both of read and write. Ignore stderr message/no block' do
+    m = Mutex.new
+    ary = []
+    Timeout.timeout(TEST_DEADLOCK_TIMEOUT) do
+      ran = false
+      # lots of stderr message should not be blocked and message should not be printed in test.
+      cmd = "ruby -e 'while !STDIN.eof? && line = STDIN.readline; STDERR.puts line.chomp * 1000; STDOUT.puts line.chomp; STDOUT.flush rescue nil; end'"
```
Comment (ganmacs) [inline test/plugin_helper/test_child_process.rb]: I'm not sure but it needs to call`STDERR.flush`?

File: test/plugin_helper/test_child_process.rb [hunk 2]
```diff
@@ -124,6 +124,39 @@ def configure(conf)
...
+        writeio.close
+
+        while line = readio.readline
+          ary << line
+        end
+        m.unlock
+      end
+      sleep TEST_WAIT_INTERVAL_FOR_BLOCK_RUNNING until m.locked? || ran
+      m.lock
+      m.unlock
```
Comment (ganmacs) [inline test/plugin_helper/test_child_process.rb]: Mutex#unlock should be called in `ensure` context for safe.

File: lib/fluent/plugin_helper/child_process.rb [hunk 3]
```diff
@@ -281,7 +282,7 @@ def child_process_execute_once(
           stderrio.set_encoding(external_encoding, internal_encoding, encoding_options)
           stderrio_in_use = true
         else
-          stderrio.reopen(IO::NULL) if stderrio && stderrio == :discard
```
Comment (ganmacs) [inline lib/fluent/plugin_helper/child_process.rb]: Just checking, This was a bug, right? `stderrio` cannot be a `:discard`.

File: test/plugin_helper/test_child_process.rb (hunk 1 above)
Comment (repeatedly) [reply test/plugin_helper/test_child_process.rb]: STDERR writes lots of data so no flush call is not critical.

File: test/plugin_helper/test_child_process.rb (hunk 2 above)
Comment (repeatedly) [reply test/plugin_helper/test_child_process.rb]: good catch.

File: lib/fluent/plugin_helper/child_process.rb (hunk 3 above)
Comment (repeatedly) [reply lib/fluent/plugin_helper/child_process.rb]: Yes. This is typo.

**Pull request merged**